File: src/main/java/net/slipp/domain/qna/ScoreLike.java
package net.slipp.domain.qna;

import javax.persistence.Column;
import javax.persistence.Entity;
import javax.persistence.EnumType;
import javax.persistence.Enumerated;
import javax.persistence.GeneratedValue;
import javax.persistence.GenerationType;
import javax.persistence.Id;
import javax.persistence.Table;

@Entity
@Table(name = "score_like")
public class ScoreLike {
    @Id
    @Column(name = "id")
    @GeneratedValue(strategy = GenerationType.AUTO)
    private Long id;

    @Enumerated(EnumType.STRING)
    @Column(name = "like_type", nullable = false, columnDefinition = ScoreLikeType.COLUMN_DEFINITION)
    private ScoreLikeType scoreLikeType;

    @Column(name = "social_user_id")
    private Long socialUserId;

    @Column(name = "target_id")
    private Long targetId;

    @Column(name = "liked")
    private boolean liked;

    public ScoreLike() {
    }
    
    private ScoreLike(ScoreLikeType scoreLikeType, Long socialUserId, Long targetId, boolean liked) {
        this.scoreLikeType = scoreLikeType;
        this.socialUserId = socialUserId;
        this.targetId = targetId;
        this.liked = liked;
    }

    public Long getId() {
        return id;
    }

    public Long getSocialUserId() {
        return socialUserId;
    }

    public ScoreLikeType getScoreLikeType() {
        return scoreLikeType;
    }

    public Long getTargetId() {
        return targetId;
    }

    public boolean isLiked() {
        return liked;
    }
    
    public static ScoreLike createLikedScoreLike(ScoreLikeType scoreLikeType, Long socialUserId, Long targetId) {
        return new ScoreLike(scoreLikeType, socialUserId, targetId, true);
    }
    
    public static ScoreLike createDisLikedScoreLike(ScoreLikeType scoreLikeType, Long socialUserId, Long targetId) {
        return new ScoreLike(scoreLikeType, socialUserId, targetId, false);
    }

    @Override
    public String toString() {
        return "ScoreLike [id=" + id + ", scoreLikeType=" + scoreLikeType + ", socialUserId=" + socialUserId
                + ", targetId=" + targetId + ", liked=" + liked + "]";
    }
}

File: src/main/scala/net/slipp/domain/qna/ScoreLike.scala
package net.slipp.domain.qna

import javax.persistence.Column
import javax.persistence.Entity
import javax.persistence.EnumType
import javax.persistence.Enumerated
import javax.persistence.GeneratedValue
import javax.persistence.GenerationType
import javax.persistence.Id
import javax.persistence.Table

object ScoreLike {
  def createLikedScoreLike(scoreLikeType: ScoreLikeType, socialUserId: Long, targetId: Long): ScoreLike = {
    new ScoreLike(scoreLikeType, socialUserId, targetId, true)
  }

  def createDisLikedScoreLike(scoreLikeType: ScoreLikeType, socialUserId: Long, targetId: Long): ScoreLike = {
    new ScoreLike(scoreLikeType, socialUserId, targetId, false)
  }
}

@Entity
@Table(name = "score_like")
class ScoreLike(likeType: ScoreLikeType, userId: Long, tId: Long, l: Boolean) {
  @Id
  @Column(name = "id")
  @GeneratedValue(strategy = GenerationType.AUTO)
  private var id: Long = _

  @Enumerated(EnumType.STRING)
  @Column(name = "like_type", nullable = false, columnDefinition = "enum('ANSWER','QUESTION')")
  private var scoreLikeType: ScoreLikeType = likeType

  @Column(name = "social_user_id")
  private var socialUserId: Long = userId

  @Column(name = "target_id")
  private var targetId: Long = tId

  @Column(name = "liked")
  private var liked: Boolean = l

  def this() = this(ScoreLikeType.QUESTION, 0L, 0L, false)

  def getId: Long = {
    return id
  }

  def getSocialUserId: Long = {
    return socialUserId
  }

  def getScoreLikeType: ScoreLikeType = {
    return scoreLikeType
  }

  def getTargetId: Long = {
    return targetId
  }

  def isLiked: Boolean = {
    return liked
  }

  override def toString: String = {
    return "ScoreLike [id=" + id + ", scoreLikeType=" + scoreLikeType + ", socialUserId=" + socialUserId + ", targetId=" + targetId + ", liked=" + liked + "]"
  }
}
